src/sinks/pulsar/sink.rs

```diff
@@ -0,0 +1,163 @@
+use bytes::Bytes;
+use std::collections::HashMap;
+
+use crate::sinks::pulsar::config::PulsarSinkConfig;
+use crate::sinks::pulsar::encoder::PulsarEncoder;
+use crate::sinks::pulsar::request_builder::PulsarRequestBuilder;
+use crate::sinks::pulsar::service::{PulsarRetryLogic, PulsarService};
+use crate::sinks::pulsar::util;
+use crate::sinks::util::{
+    ServiceBuilderExt, SinkBuilderExt, TowerRequestConfig, TowerRequestSettings,
+};
+use crate::template::{Template, TemplateParseError};
+use crate::{
+    codecs::{Encoder, Transformer},
+    event::Event,
+};
+use futures::{stream::BoxStream, StreamExt};
+use pulsar::{Error as PulsarError, Pulsar, TokioExecutor};
+use serde::Serialize;
+use snafu::{ResultExt, Snafu};
+use tower::ServiceBuilder;
+use vector_buffers::EventCount;
+use vector_common::byte_size_of::ByteSizeOf;
+use vector_core::{
+    event::{EstimatedJsonEncodedSizeOf, LogEvent},
+    sink::StreamSink,
+};
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub(crate)))]
+pub(crate) enum BuildError {
+    #[snafu(display("creating pulsar producer failed: {}", source))]
+    CreatePulsarSink { source: PulsarError },
+    #[snafu(display("invalid topic template: {}", source))]
+    TopicTemplate { source: TemplateParseError },
+}
+
+pub(crate) struct PulsarSink {
+    transformer: Transformer,
+    encoder: Encoder<()>,
+    service: PulsarService<TokioExecutor>,
+    request_settings: TowerRequestSettings,
+    config: PulsarSinkConfig,
+    topic_template: Template,
+}
+
+/// Stores the event together with the extracted keys, topics, etc
+/// This is passed into the `RequestBuilder` which then splits it out into the event
+/// and metadata containing the keys, and metadata
+/// This event needs to be created prior to building the request so we can filter out
+/// any events that error whilst rendering the templates.
+#[derive(Serialize)]
+pub(super) struct PulsarEvent {
+    pub(super) event: Event,
+    pub(super) topic: String,
+    pub(super) key: Option<Bytes>,
+    pub(super) properties: Option<HashMap<String, Bytes>>,
+    pub(super) timestamp_millis: Option<i64>,
+}
+
+impl EventCount for PulsarEvent {
+    fn event_count(&self) -> usize {
+        // A PulsarEvent represents one event.
+        1
+    }
+}
+
+impl ByteSizeOf for PulsarEvent {
+    fn allocated_bytes(&self) -> usize {
+        self.event.size_of()
+            + self.topic.size_of()
+            + self.key.as_ref().map_or(0, |bytes| bytes.clone().size_of())
+            + self.properties.as_ref().map_or(0, |props| {
+                props
+                    .iter()
+                    .map(|(key, val)| key.capacity() + val.size_of())
+                    .sum()
+            })
+    }
+}
+
+impl EstimatedJsonEncodedSizeOf for PulsarEvent {
+    fn estimated_json_encoded_size_of(&self) -> usize {
+        self.event.estimated_json_encoded_size_of()
+    }
+}
+
+pub(crate) async fn healthcheck(config: PulsarSinkConfig) -> crate::Result<()> {
```

Comment:
makes sense to me, i'll move them

Reply:
Again credit to @bruceg - feels like a nice QOL improvement for debugging, might be worth a grep or two to see if we're emitting logging for another healthcheck - not that it would block the change (IMHO).